dht.go

```diff
@@ -89,17 +90,53 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
 	if err := cfg.Apply(append([]opts.Option{opts.Defaults}, options...)...); err != nil {
 		return nil, err
 	}
-	dht := makeDHT(ctx, h, cfg.Datastore, cfg.Protocols, cfg.BucketSize)
 
-	// register for network notifs.
-	dht.host.Network().Notify((*netNotifiee)(dht))
+	// set seeder, snapshotter & fallback peers if not set
+	if cfg.Persistence.Seeder == nil {
+		cfg.Persistence.Seeder = persist.NewRandomSeeder(h, persist.DefaultRndSeederTarget)
+	}
+	if cfg.Persistence.Snapshotter == nil {
+		s, err := persist.NewDatastoreSnapshotter(cfg.Datastore, persist.DefaultSnapshotNS)
+		// should never happen
+		if err != nil {
+			logger.Error("failed to initialize the default datastore backed snapshotter")
+			panic(err)
+		}
+		cfg.Persistence.Snapshotter = s
+	}
+	if len(cfg.Persistence.FallbackPeers) == 0 {
+		cfg.Persistence.FallbackPeers = DefaultBootstrapPeerIDs
+	}
+
+	dht := makeDHT(ctx, h, &cfg, cfg.BucketSize)
 
 	dht.proc = goprocessctx.WithContextAndTeardown(ctx, func() error {
 		// remove ourselves from network notifs.
 		dht.host.Network().StopNotify((*netNotifiee)(dht))
 		return nil
 	})
 
+	// fetch the last snapshot & feed it to the seeder
+	candidates, err := cfg.Persistence.Snapshotter.Load()
+	if err != nil {
+		logger.Warningf("error while loading snapshot of DHT routing table: %s, cannot seed dht", err)
+	} else if err := cfg.Persistence.Seeder.Seed(dht.routingTable, candidates, cfg.Persistence.FallbackPeers); err != nil {
+		logger.Warningf("error while seeding candidates to the routing table: %s", err)
+	}
+
+	// schedule periodic snapshots
+	sproc := periodicproc.Tick(cfg.Persistence.SnapshotInterval, func(proc goprocess.Process) {
```

Comment:
I see what you mean. So, in it's current form, this code could persist an empty RT if the snapshotting go-routine fires after all peers have been dropped from the RT, but Dht hasn't been closed yet.

However, this can also prove to be a blessing in disguise because storing an empty RT & then seeding with bootstrap peers after we restart could save us from ALWAYS storing a poisoned RT if an attacker messed up our RT and found a way to immediately shut us down.

So, let's see how this works in practise & fix it if required ?